### PR TITLE
WIP: Resolve losing keyboard shortcuts on login

### LINKED ISSUE #1763: Some Media Keys (Not all) and custom keyboard shortcuts irregulary don't work after boot/reboot
Copy paste  from the launchpad bug report.  Note - we have had several reports over the years but this bug report has been analysed by a kind GNOME dev and gave a possible implication of what is causing the issue.

https://bugs.launchpad.net/ubuntubudgie/+bug/1818978

## Description
After boot or reboot some media keys (Volume Up/Down) don't work, but others do (e.g. Brightness Up/Down). Custom Keyboard Shortcuts then don't work at all. When a non working media key oder shortcut is pressed the following message appears in syslog:

gsd-media-keys[2016]: Could not find accelerator for accel id 169

This happens irregularly. Sometimes everything works, sometimes it doesn't work as described above until next reboot.

This bug was first reported here: https://gitlab.gnome.org/GNOME/gnome-settings-daemon/issues/151 (And also here: https://bugs.launchpad.net/ubuntu/+source/gnome-settings-daemon/+bug/1817286 but important information is provided by first link)
Benjamin Berg, Maintainer of gnome-settings-daemon, found out it's related to budgies implementation of gsd-media-keys. Please look at this issue for logs and details. He even pointed out where the root issue might be (It seems like 'UngrabAccelerator' isn't implemented correctly)

I can only reproduce this bug inside budgie desktop enviornment, not inside an ubuntu gnome session installed via 'sudo apt-get install ubuntu-gnome-desktop'

## Budgie version
budgie desktop v10.4 and 10.5

## Operating system
Ubuntu Budgie 18.04/18.10/19.04

## Steps to reproduce the issue
This is really hard to reproduce - it does appear to be random - but mainly after a clean boot/reboot rather than a logout and login

"After boot or reboot some media keys (Volume Up/Down) don't work, but others do (e.g. Brightnes Up/Down). Custom Keyboard Shortcuts then don't work at all."

## Actual result
some media keys (Volume Up/Down) don't work, but others do

## Expected result
All media keys should work

## Additional information
Linked GNOME report https://gitlab.gnome.org/GNOME/gnome-settings-daemon/issues/151

"Please point your distribution/budgie developers to:

Mär 05 17:07:52 BaBook budgie-wm[11980]: g_hash_table_iter_next: assertion 'ri->version == ri->hash_table->version' failed

https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#g-hash-table-iter-init saying Modifying the hash table after calling this function invalidates the returned iterator

https://developer.gnome.org/glib/stable/glib-Hash-Tables.html#g-hash-table-iter-remove
https://github.com/solus-project/budgie-desktop/blob/master/src/wm/shim.vala#L222

 Benjamin Berg @bberg · 1 day ago
Maintainer

Hah, I looked a bit more. Budgie doesn't implement UngrabAccelerator …
"

EDIT:

Currently testing whether changing https://github.com/solus-project/budgie-desktop/blob/master/src/wm/shim.vala#L222 to `iter.remove()` resolves this issue.

### PR DESCRIPTION
**Pull request by fossfreedom**
## Description
WIP: Resolves #1763

From #1763 linked issue to mutter tracker, the maintainer their had concerns how budgie desktop was managing keyboard dbus calls to grabbing and ungrabbing shortcuts.

I've reworked this along the guidelines the maintainer requested.

Marked this as WIP because it does need much wider testing because of the irregular nature of the issue.

### Submitter Checklist

- [X] Squashed commits with `git rebase -i` (if needed)
- [X] Built budgie-desktop and verified that the patch worked (if needed)

### CODE REVIEW

**Comment by vredesbyyrd** [comment]
I have been testing this on my daily driver laptop for the last week. I'm running arch + budgie and prior to building with this commit I would lose custom shortcuts often. With many logouts and restarts I have not run into said bug yet. 

Cheers.

File: src/wm/shim.vala [hunk 1]
```diff
@@ -196,60 +194,58 @@ public class ShellShim : GLib.Object
...
-        params.insert("device-id", device_id);
-        params.insert("timestamp", new Variant.uint32(0));
-        params.insert("action-mode", new Variant.uint32(0));
-
-        this.accelerator_activated(action, params);
+        foreach (string accelerator in grabs.get_keys ()) {
+            if (grabs[accelerator] == action) {
+                var params = new GLib.HashTable<string, Variant> (null, null);
+                params.set ("device-id", new Variant.uint32 (device_id));
+#if HAV_MUTTER_5
```
Comment (JoshStrobl) [inline src/wm/shim.vala]: It's never going to set this as the timestamp because this should be HAVE_MUTTER_5

**Comment by fossfreedom** [comment]
@vredesbyyrd (and anyone else interested) I have tidied up the patch.  It was missing one specific dbus call implementation for mutter 5.  I would appreciate that you grab the latest changes and give it a test over the next week or so to confirm no adverse affects.  TIA

**Comment by vredesbyyrd** [comment]
@fossfreedom 

I intended to comment sooner.

On your first iteration of this patch I did run into broken custom keybindings a couple times over the last month. Not nearly as frequently as without the patch. 

I have been testing commit b812847a898193d475ab68738d78bfb75b6a3692 for a few days now, so far without issue. 

>  It was missing one specific dbus call implementation for mutter 5...

Sounds like you resolved this, but if said bug pops up again over the coming days ill get back. Thanks for the patch!

**Pull request merged**